### PR TITLE
Consider remediations for packages' vulnerabilities scan

### LINKED ISSUE #23066: Packages vulnerability remediations do not work
### Description
This issue aims to add the capacity to detect remediations in CVEs during the package scanner.

For example, some Windows applications have associated some hotfixes.
Ex: Microsoft Office 2016 and 2019 have the associated CVE-2024-21413 that have the "KB5002467", "KB5002519", "KB5002522", "KB5002537" as remediations.

# Current Behavior
- CVE-2024-21413 is marked as vulnerable.

# Expected
- CVE-2024-21413 is discarded because you have some of the "KB5002467", "KB5002519", "KB5002522", "KB5002537"

# Steps
- Design. (1 day) @Dwordcito @sebasfalcone   
- Implementation. (2 days) @sebasfalcone 
- Development test. (1 day) @GabrielEValenzuela  
- Review. (1 day) @wazuh/devel-core2-div1

### PR DESCRIPTION
**Pull request by pereyra-m**
|Related issue|
|---|
|Closes #23066|

## Description

This PR solves two different situations:
- The Windows packages that have hotfixes report false positives
- The insertion of a hotfix solves a package vulnerability

It was required to create a new cache with the agents' hotfixes and to store all the CVEs that affect a hotfix in a new DB column.

## Logs/Alerts example

The vulnerability shown in the issue isn't being reported because the versions don't match.
After inserting a temporal candidate with the version `16.0.4266.1001`

```
./src/build/wazuh_modules/vulnerability_scanner/testtool/rocksDBQuery/rocks_db_query_testtool -d /var/ossec/queue/vd/feed/ -f /workspaces/wazuh/src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityCandidate.fbs -c nvd -k office_CVE-2024-21413 -v '{  "candidates": [    {      "cveId": "CVE-2024-21413",      "defaultStatus": "unaffected",      "versions": [        {          "version": "2016"        },        {          "version": "2019"        },        {          "version": "16.0.4266.1001"        }      ],      "vendor": "microsoft"    }  ]}'
```

The vulnerability is found

<details><summary>Details</summary>
<p>

```
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:353 at vendorVerify(): DEBUG: Vendor match for Package: office, Version: 16.0.4266.1001, CVE: CVE-2024-21413, Vendor: microsoft
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:386 at versionMatch(): DEBUG: Scanning package - 'office' (Installed Version: 16.0.4266.1001, Security Vulnerability: CVE-2024-21413). Identified vulnerability: Version: 2016. Required Version Threshold: . Required Version Threshold (or Equal): .
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:386 at versionMatch(): DEBUG: Scanning package - 'office' (Installed Version: 16.0.4266.1001, Security Vulnerability: CVE-2024-21413). Identified vulnerability: Version: 2019. Required Version Threshold: . Required Version Threshold (or Equal): .
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:386 at versionMatch(): DEBUG: Scanning package - 'office' (Installed Version: 16.0.4266.1001, Security Vulnerability: CVE-2024-21413). Identified vulnerability: Version: 16.0.4266.1001. Required Version Threshold: . Required Version Threshold (or Equal): .
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:406 at versionMatch(): DEBUG: Match found, the package 'office', is vulnerable to 'CVE-2024-21413'. Current version: '16.0.4266.1001' is equal to '16.0.4266.1001'. - Agent 'DESKTOP-EQ4F57D' (ID: '001', Version: 'v4.7.4').
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:583 at packageHotfixSolved(): DEBUG: No remediation for package 'office' on agent '001' that solves CVE 'CVE-2024-21413' has been found.
2024/05/03 04:02:57 wazuh-modulesd:vulnerability-scanner[164578] packageScanner.hpp:705 at handleRequest(): DEBUG: Vulnerability scan for package 'Microsoft Office Professional Plus 2016' on Agent '001' has completed.
```

</p>
</details> 

If we repeat the scan but after installing a hotfix that solves the CVE, it isn't reported as vulnerable

<details><summary>Details</summary>
<p>

```
2024/05/03 14:53:38 wazuh-modulesd:vulnerability-scanner[19994] packageScanner.hpp:387 at versionMatch(): DEBUG: Scanning package - 'office' (Installed Version: 16.0.4266.1001, Security Vulnerability: CVE-2024-21413). Identified vulnerability: Version: 16.0.4266.1001. Required Version Threshold: . Required Version Threshold (or Equal): .
2024/05/03 14:53:38 wazuh-modulesd:vulnerability-scanner[19994] packageScanner.hpp:407 at versionMatch(): DEBUG: Match found, the package 'office', is vulnerable to 'CVE-2024-21413'. Current version: '16.0.4266.1001' is equal to '16.0.4266.1001'. - Agent 'DESKTOP-EQ4F57D' (ID: '001', Version: 'v4.7.4').
2024/05/03 14:53:38 wazuh-modulesd:vulnerability-scanner[19994] packageScanner.hpp:571 at packageHotfixSolved(): DEBUG: Remediation 'KB5002537' for package 'office' on agent '001' that solves CVE 'CVE-2024-21413' has been found.
2024/05/03 14:53:38 wazuh-modulesd:vulnerability-scanner[19994] packageScanner.hpp:706 at handleRequest(): DEBUG: Vulnerability scan for package 'Microsoft Office Professional Plus 2016' on Agent '001' has completed.
```

</p>
</details> 

If we now run the scan without the hotfix and the install it, we get the alert of vulnerability solved

WIP

## Tests
- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Review logs syntax and correct language
- [x] Added unit tests (for new features)

### CODE REVIEW

**Comment by MiguelazoDS** [comment]
# Testing 

## Environment 

- Manager Ubuntu Jammy
- Agent Centos9 
- Agent Windows 11

## Scan by events 

- Total vulnerabilities found 

[vulnerabilities.json](https://github.com/wazuh/wazuh/files/15266044/vulnerabilities.json)
[ossec.log.tar.gz](https://github.com/wazuh/wazuh/files/15266138/ossec.log.tar.gz)

> [!WARNING]
> Even though the logs are at the debug level, the file increases considerably. 

image
image

78 vulnerabilities for Ubuntu Jammy
2139 vulnerabilities for Centos 9 
177 vulnerabilities for Windows 11 
image

> [!WARNING]
> All vulnerabilities have the "Packages" category

image

100 vulnerabilities are related to Office 

image

## Scan after installing new KB

**Comment by sebasfalcone** [comment]
## Testing - Install a KB for a vulnerable package 🟢 
- Installed Office 2016
image

- Installed KB to fix
```
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_0011be200caf61ef8d0d0477e4e8f328c2c016ae_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_094055ceb0e2e63b0ecfcbb41a66541a1b1ca083_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_222ae3df7ad2779d4a42a9b02514089740cef63b_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_308b309694f41a18b3f99e1c8b05201eeafca00c_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_9a73b12da93578701cd1fab0b334407cf33d2dc0_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_9c5af1826923047935b52622447348b5b8160da4_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_aa43ca9f0c8b334c4f8270b0649dc05356bf9f7c_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_d9287b970add8f202e3f3e84012d6431ba79a471_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_e04aea6001b3db2555f84bae073b6e51a45636ff_CVE-2024-21413'.
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] cveSolvedInventorySync.hpp:72 at operator()(): DEBUG: CVE 'CVE-2024-21413' was remediated by hotfix 'KB5002467' for '002_efdf9aa5c8664cedd578be67fc58dd43119e2b30_CVE-2024-21413'.


2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] eventSendReport.hpp:89 at handleRequest(): DEBUG: Vulnerability report for agent ID 002, hotfix: KB5002467, cve: CVE-2024-21413


2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413
2024/05/09 17:36:00 wazuh-modulesd:vulnerability-scanner[245855] resultIndexer.hpp:56 at operator()(): DEBUG: Processing and publish key: CVE-2024-21413

```
image

**Comment by sebasfalcone** [comment]
## Test - Install vulnerable package (KB already installed)
- Initial status
image

- Install Office 2016
image

- CVE-2024-21413 was not detected, as it was fixed with the KB

**Review by MiguelazoDS — APPROVED** [review]
GJ! functional testing succeeded.

**Review by MiguelazoDS — DISMISSED** [review]

**Review by Dwordcito — CHANGES_REQUESTED** [review]
Code looks good, minor things. I'm worried that we didn't add a single component qa test, this shouldn't be too difficult to test with the testing framework we use.

**Comment by Dwordcito** [comment]
please, rebase this class.

**Comment by GabrielEValenzuela** [comment]
# Testing 🟡 

## Environment

- Manager 4.8.0 on Ubuntu 22
- Agents: 
  -  Windows 10 x 2
  - Windows 11
  - Windows 8.1
  - Windows 7
  - Alpine
  - Debian12
  - Centos 8
  - Rocky 8
  - RedHat 7
  - AlmaLinux 9
  - Amazon 2023
  - Ubuntu 22
  - Ubuntu 20
  - SLE15
- Database provided from drive

## Results (Brief)

- Some packages were not scanned, for example, manager and SLE15
- The Windows 10 agent has vulnerabilities of OS type along with packages like Edge, after updating with the last KBs no solved solution was found. Also connecting a second agent doesn't show the expected logs.

```bash
2024/05/14 16:47:27 wazuh-modulesd:vulnerability-scanner[75886] hotfixInsert.hpp:67 at handleRequest(): DEBUG: Getting associated vulnerabilities for hotfix 'KB5033052'
2024/05/14 16:47:27 wazuh-modulesd:vulnerability-scanner[75886] hotfixInsert.hpp:73 at handleRequest(): DEBUG: No vulnerabilities associated to hotfix 'KB5033052'
```

- Only Windows 8.1 generate alerts
- Compared with the previous scan, the scan of Ubuntu packages was different (Few vulnerabilities found)
- The amount of vulnerable packages in the first scan differs with the rescan (More in the re-scan)

### Log attached

> [!NOTE]
> Remove the zip extension

[logs.tar.xz.zip](https://github.com/wazuh/wazuh/files/15313280/logs.tar.xz.zip)

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/remediationDataCache.hpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+{
+    std::unordered_set<std::string> hotfixes; ///< Installed hotfixes.
+};
+
+/**
+ * @brief remediationDataCache class.
+ *
+ * @note This class queries the Wazuh-DB to get the remediation data for a given agent, and stores it in a LRU cache
+ */
+class remediationDataCache final : public Singleton<remediationDataCache>
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/remediationDataCache.hpp:43]: Done

File: src/shared_modules/utils/socketDBWrapper.hpp [hunk 2]
```diff
@@ -56,10 +56,10 @@ class SocketDBWrapper final : public Singleton<SocketDBWrapper>
     bool m_dataReady {false};
 
 public:
-    void init()
+    void init(const std::string& socketPath = WDB_SOCKET)
     {
         m_teardown = false;
-        m_dbSocket = std::make_unique<SocketClient<Socket<OSPrimitives, SizeHeaderProtocol>, EpollWrapper>>(WDB_SOCKET);
```
Comment (Dwordcito) [inline src/shared_modules/utils/socketDBWrapper.hpp:62]: Why is the socket path parameter if this is only for WDB?

File: src/shared_modules/utils/socketDBWrapper.hpp [hunk 3]
```diff
@@ -56,10 +56,10 @@ class SocketDBWrapper final : public Singleton<SocketDBWrapper>
     bool m_dataReady {false};
 
 public:
-    void init()
+    void init(const std::string& socketPath = WDB_SOCKET)
```
Comment (Dwordcito) [inline src/shared_modules/utils/socketDBWrapper.hpp:59]: Why is the socket path parameter if this is only for WDB?

Comment (sebasfalcone) [reply src/shared_modules/utils/socketDBWrapper.hpp:59]: This make the use of the testtool not possible, we can revert this changes

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp [hunk 4]
```diff
@@ -1397,6 +1416,72 @@ struct TScanContext final
...
+            []([[maybe_unused]] const nlohmann::json* jsonData) { return ""; });
+    }
+
+    /**
+     * @brief Checks if there are remediation installed.
+     *
+     * @return true if there are remediation.
+     * @return false if there are not remediation.
+     */
+    bool hasRemediations() const
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:1458]: Removed

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp [hunk 5]
```diff
@@ -1397,6 +1416,72 @@ struct TScanContext final
...
+
+    /**
+     * @brief Checks if a remediation is installed.
+     *
+     * @param remediation ID of the remediation to check.
+     *
+     * @return true if the remediation is installed.
+     * @return false if the remediation is not installed.
+     */
+    bool remediationIsInstalled(const std::string& remediation) const
```
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/scanContext.hpp:1471]: Removed

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/resultIndexer.hpp [hunk 6]
```diff
@@ -54,14 +54,7 @@ class TResultIndexer final : public AbstractHandler<std::shared_ptr<TScanContext
...
                 logDebug2(WM_VULNSCAN_LOGTAG, "Processing and publish key: %s", key.c_str());
-                if (value.contains("operation") && value.contains("id"))
-                {
-                    m_indexerConnector->publish(value.dump());
-                }
-                else
-                {
-                    logError(WM_VULNSCAN_LOGTAG, "Invalid element to publish: %s", value.dump().c_str());
-                }
+                m_indexerConnector->publish(value.dump());
```
Comment (Dwordcito) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/resultIndexer.hpp:57]: Why remove the contains for operation and id ?

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/arrayResultIndexer.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                {
+                    for (const auto& element : value)
+                    {
+                        resultCallback(element, key);
+                    }
+                }
+                else
+                {
+                    resultCallback(value, key);
+                }
```
Comment (Dwordcito) [inline src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/arrayResultIndexer.hpp:79]: If it is just an array, we should not do anything if it is an object or throw an exception.

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/resultIndexer.hpp (hunk 6 above)
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/resultIndexer.hpp:57]: A poor job restoring the changes

File: src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/arrayResultIndexer.hpp (hunk 7 above)
Comment (sebasfalcone) [reply src/wazuh_modules/vulnerability_scanner/src/scanOrchestrator/arrayResultIndexer.hpp]: Added handling and UTs

**Comment by pereyra-m** [comment]
## Update 

The QA tests for vulnerability scanner fail because the content has changed and it still hasn't been published.
The tests pass in a local environment

**Review by Dwordcito — DISMISSED** [review]

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**